### DISCUSSION TITLE
Enhance SOFA with robotics support

### DISCUSSION THREAD

**Discussion by HoangGiang93** [sofa-framework/sofa#3898] [Ideas / Suggestions]
This proposal is intended to enhance SOFA with features that facilitate robotic simulation. The objective is to enable SOFA to simulate a robot performing tasks with various physical objects in a virtual setting. To attain this, SOFA needs to incorporate the following capabilities:

1. Ability to import a robot model from URDF.

2. Incorporate both ROS1 and ROS2 support (despite ROS1's anticipated deprecation in 2026, its wide usage makes this feature relevant), and include a ros_control interface.

3. Employ a mechanism to convert a scene description into a SOFA-compatible format. While there are numerous scene descriptions such as SDF in Gazebo, MJCF in MuJoCo, PROTO in Webot, and USD in Isaac Sim/Omniverse, it is highly recommended to develop a parser for USD. This is because USD can be extended to be convertible from any other formats, and it includes a built-in python binding that facilitates efficient reading and writing of USD files.

4. Facilitate the spawning and destruction of objects in real-time. These objects can be either rigid or soft, a simple primitive with basic properties (like a cube with mass, inertia, and size), or an object stored in static memory (like a cup or a sponge). The spawning and destruction event can be triggered by a ROS service.

Any feedback on these ideas would be great @hugtalbot @olivier-roussel

**Comment by HoangGiang93** [comment]
Do you have any update on this @olivier-roussel ?

**Comment by bakpaul** [comment]
Hi Hoang, I may have a begining of an answer for your regarding your second point : Ros2 support. 
An opensource plugin is already available for use merging Ros2 and Sofa, you can download it here : https://github.com/courtecuisse/sofaros2plugin

For the rest I'll let the one actually working on  robotic in sofa answer you.

**Comment by olivier-roussel** [comment]
Hi @HoangGiang93,
Sorry I am just discovering your message. I probably missed the notification and as I am not a full-time SOFA developper, I was mostly working on other stuff these times.
I still plan to integrate Pinocchio software into SOFA through a future plugin. The development should start soon and even if it will require a bit of time, it may answer quickly to some of your points, i.e. :
1. Import robot model from URDF: This is planned to be part of the plugin, as Pinocchio embeds a URDF parser.
2. ROS / ROS2 support: I did not had a look to it yet but the answer of @bakpaul looks like a good start for a ROS2 plugin. Let us know your feedback on this. For the ros_control interface, if it is not existant yet, that will be probably something I will need too so I am also interested.
3. The SDFormat is planned to be fully integrated on the next Pinocchio release (3.x), which will be used for the robotics plugin I will develop. For USD or other formats, I don't know them. This can be investigated in the future, but are not planned for the first release.
4. Isn't it already possible for soft bodies ? Like through the python interface for example.

**Reply by HoangGiang93** [reply]
Hi @olivier-roussel,
Thank you for your answer. 1) and 2) are clear for me and I'm happy to wait for that. The SDFormat in 3) could be useful for importing / exporting scene description, makes it more convenient to launch the simulation. I also developed a parser to convert from SDF to USD and vice versa, so that would be no problem.
4) Yes it's possible, but it's still hard to use because one should know how to tune the proper parameters to have a feasible simulator for a simple case. I think a simple service that does everything for us with minimal interface and could be trigger by a script would be nice. For instance, if I want to spawn a cube with size 0.1m on each side, I don't want to specify the others parameters, just spawn a cube with some default parameters.

**Comment by alxbilger** [comment]
Since it has not been mentioned before, I put that here: https://github.com/SofaDefrost/SoftRobots/tree/master/examples/sofaros
